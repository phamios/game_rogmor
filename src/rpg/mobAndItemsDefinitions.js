export const SEX = {
  male: 'male',
  female: 'female',
};

export const mobDefinitions = [
  {title: "", descript: "elf girl", sex: SEX.female },
  {title: "", descript: "box face", sex: SEX.male },
  {title: "", descript: "chaotic hair", sex: SEX.male },
  {title: "", descript: "bold face", sex: SEX.male },
  {title: "", descript: "fat face", sex: SEX.male },
  {title: "", descript: "charming knight", sex: SEX.male },
  {title: "", descript: "elf merchant", sex: SEX.male },
  {title: "", descript: "cow one", sex: SEX.female },
  {title: "", descript: "elf bard", sex: SEX.male },
  {title: "", descript: "elf girl with dark hair", sex: SEX.female },
  {title: "", descript: "young troll", sex: SEX.male },
  {title: "", descript: "horned inspector", sex: SEX.male },
  {title: "", descript: "citizen", sex: SEX.male },
  {title: "", descript: "blond hair", sex: SEX.female },
  {title: "", descript: "pig one", sex: SEX.male },
  {title: "", descript: "sqared orc", sex: SEX.male },
  {title: "", descript: "bar tender ogre", sex: SEX.male },
  {title: "", descript: "wizard", sex: SEX.male },
  {title: "", descript: "handsome human", sex: SEX.male },
  {title: "", descript: "neighbour", sex: SEX.male },
  {title: "", descript: "long hair", sex: SEX.male },
  {title: "", descript: "six horn", sex: SEX.male },
  {title: "", descript: "desert warrior", sex: SEX.male },
  {title: "", descript: "sad fighter", sex: SEX.male },
  {title: "", descript: "trader", sex: SEX.male },
  {title: "", descript: "paladin", sex: SEX.male },
  {title: "", descript: "knight girl", sex: SEX.female },
  {title: "", descript: "adventurer", sex: SEX.male },
  {title: "", descript: "psycho", sex: SEX.male },
];

export const itemDefinitions = [
  {title: "", descript: "saw knife"},
  {title: "", descript: "short sword"},
  {title: "", descript: "halberd"},
  {title: "", descript: "bow"},
  {title: "", descript: "iron shield"},
  {title: "", descript: "rounded shield"},
  {title: "", descript: "sword"},
  {title: "", descript: "dagger"},
  {title: "", descript: "spiked jerk"},
  {title: "", descript: "staff"},
  {title: "", descript: "bishop staff"},
  {title: "", descript: "axe"},
  {title: "", descript: "ring"},
  {title: "", descript: "bold blade"},
  {title: "", descript: "rapier"},
  {title: "", descript: "hammer"},
  {title: "", descript: "recoil bow"},
  {title: "", descript: "xiphos"},
  {title: "", descript: "harpoon"},
  {title: "", descript: "spear"},
  {title: "", descript: "great shield"},
  {title: "", descript: "knife"},
  {title: "", descript: "mallet"},
  {title: "", descript: "long sword"},
  {title: "", descript: "dual axe"},
  {title: "", descript: "sabre"},
  {title: "", descript: "knight sword"},
  {title: "", descript: "crossbow"},
  {title: "", descript: "claw"},
  {title: "", descript: "broad sword"},
  {title: "", descript: "barbarian axe"},
  {title: "", descript: "saw dagger"},
  {title: "", descript: "cup"},
  {title: "", descript: "material"},
  {title: "", descript: "chest"},
  {title: "", descript: "bag"},
  {title: "", descript: "coin"},
  {title: "", descript: "anvil"},
  {title: "", descript: "box"},
  {title: "", descript: "potion"},
  {title: "", descript: "stone"},
  {title: "", descript: "chicken"},
  {title: "", descript: "goblet"},
  {title: "", descript: "flag"},
  {title: "", descript: "point of interest"},
];