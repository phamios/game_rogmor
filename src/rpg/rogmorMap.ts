// y * 1000 + x
export const toCoord = ({x, y}) => y * 1000 + x;
export const abToCoord = (a, b) => b * 1000 + a;
export const coordTo = coord => ({x: coord % 1000 , y: coord / 1000 | 0 }); 
export const coordToStyle = coord => ({left: (coord % 1000) * 40, top: (coord / 1000 | 0) * 40}); 
export const dryLand:number[] = [8007,8008,8009,8010,7010,6010,5010,4010,3010,3009,2009,1009,1010,1011,1012,1013,1014,2014,2013,2012,2011,2010,3011,3012,3013,3014,3015,4011,4012,4013,5013,5014,4014,4015,4016,4017,3017,3018,4018,4019,5019,6019,7019,8019,9019,9018,9017,9016,10016,11016,11015,11014,11013,12014,12015,10015,10014,9014,9015,10017,11017,8018,8017,7017,7018,6018,5018,5017,5016,5015,6015,6016,6017,7016,7015,7014,7013,7012,7011,6011,5011,5012,6012,6013,6014,8015,8014,8013,8012,8011,9011,9012,10011,10010,10009,9009,9010,9008,9007,9006,10008,10007,11007,12007,13007,13008,12006,13006,13005,13004,12004,12003,11003,11002,10002,9002,9001,8001,7001,7000,6001,6002,5002,4002,3002,3001,2001,1001,1002,2002,3003,2003,2004,3004,4004,5004,5003,4003,6003,6004,7004,7005,6005,8005,9005,9004,10004,11004,11005,11006,12005,10005,10006,10003,9003,8003,7003,7002,8002,8004,8006,7006,7007,7008,7009,6009,5009,5008,6008,6007,5007,5006,4006,4007,4008,3006,2006,2007,10013,12013,12012,12011,13011,13012,13010,12010,12009];