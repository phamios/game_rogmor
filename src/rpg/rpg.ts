export const leveling = (level:number, mod:number = 0.5, pow:number = 4, madd:number = 2) => Math.round(
  mod * Math.pow(level, pow) + (madd * level)
);

export const rnd = (dice:number) => dice * Math.random() | 0;

export const dice = (roll:number) => rnd(roll - 1) + 1;

// export const improved = ( attribute = 1, percent = 0.2 ) => (attribute + (Math.random() * percent * attribute)) | 0;
export const improved = ( attribute:number = 1 ) => attribute * 1.15 | 0;

export const shuffle = () => Math.random() > 0.5 ? 1 : -1;

export const pickOne = (arr:any[]) => arr[rnd(arr.length)];

export const selectRandomPick = (element:any, index:number, arr:any[], pickIndex = Math.random() * arr.length | 0) => index === pickIndex;

export const uid = () => Math.random().toString(32).slice(-8);

export const amount = (n:number) => Array(n).fill(0).map((_, index) => index);
