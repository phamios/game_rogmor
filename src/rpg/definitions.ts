import { Mob } from "./profession";

export interface Effect {
  id: string;
  type: string;
}
export interface Act {
  id: string;
  type: string;
  cooldown: number;
  tick: number;
  script: string;
}

export interface ActInstance {
  act: Act;
  actor: Mob;
  targetList: Mob[];
  timestamp?: number;
}

export interface Encounter {
  mobList: Mob[];
  actList: ActInstance[];
  targetList: Mob[];
  showTime: Effect[];
}

export type OrderOfSeed = [Mob, number];

export enum Outcome { ENDED };
export type OutcomeList = Outcome[];
