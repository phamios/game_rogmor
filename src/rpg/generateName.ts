import { rnd } from "./rpg";

export const ROGMOR_NAME_SOURCE = 'proedmorainowerialaenderomanulnivosudanterill';

export const generateName = (source = ROGMOR_NAME_SOURCE) => {
  const name:string = Array.from(
      {length:2 + rnd(3)}, 
      (s:number = rnd(source.length)) => source.slice(s, s + 2 + rnd(3))
    )
    .join('')
  ;
  const [capital, ...rest] = name.split('');
  return [capital.toUpperCase(),...rest].join('');
}
