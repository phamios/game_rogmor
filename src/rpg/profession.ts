import { rnd, leveling, uid, pickOne } from './rpg';

export interface Ability { 
  title:string;
  body: number;
  reflex: number;
  soul: number;
  aura: number;
  stamina: number;
  focus: number;
  morale: number; 
}

export interface Condition {
  staminaState: number;
  focusState: number;
  moraleState: number;
  isOut?: boolean;
}

export enum Team { PLAYER, GOOD, BAD, UGLY };

export interface MobTraits {  
  level: number;
  professionType: ProfessionKey;
  ability: Ability;
  condition: Condition;
}
export interface MobPerson {  
  name: string;
  avatar: number;
  coord: number;
  uid: string;
  team: Team;
}

export interface Mob extends MobTraits, MobPerson {}

export type ProfessionKey =
  'fighter' |'priest' |'paladin' |'knight' |'assasin' |'rogue' |'mage' |'captain' |'noble' |
  'headHunter' |'witch' |'valkur' |'guard' |'gladiator' |'barbarian' |'wizard' |'sage' |
  'samurai' |'ninja' |'warlord' |'bishop' |'tracker' |'dementor' |'necromanta' |'summoner' |
  'shaman' |'thief' |'rouge' |'cavalier' |'pirate' |'inquisitor' |'icelander' |'heretic' |
  'nomad' |'bard' |'warlock' |'monk' |'druid' |'illusionist' |'wardancer' |'runeSmith' |
  'merchant' |'explorer' |'templar' |'spy' |'nun' |'mercenary' |'duelist' |'beastmaster' |
  'sorcerer' |'archer' |'psionic' |'alchemist' |'enginer' |'antiPaladin' |'chaosKnight' |
  'chief' |'maruder' |'rich' |'shapechanger' 

export type ProfessionLookup = Record<ProfessionKey, Ability>; 

export const professionLookup:ProfessionLookup = {
    fighter:      { title:'Fighter',      body: 7, reflex: 4, soul: 2, aura: 4, stamina: 55, focus: 45, morale: 40 },
    priest:       { title:'Priest',       body: 3, reflex: 3, soul: 7, aura: 8, stamina: 35, focus: 65, morale: 40 },
    paladin:      { title:'Paladin',      body: 6, reflex: 3, soul: 6, aura: 6, stamina: 50, focus: 50, morale: 40 },
    knight:       { title:'Knight',       body: 7, reflex: 3, soul: 3, aura: 5, stamina: 60, focus: 50, morale: 40 },
    assasin:      { title:'Assasin',      body: 5, reflex: 7, soul: 1, aura: 3, stamina: 45, focus: 55, morale: 40 },
    rogue:        { title:'Rogue',        body: 4, reflex: 6, soul: 3, aura: 5, stamina: 55, focus: 40, morale: 40 },
    mage:         { title:'Mage',         body: 3, reflex: 4, soul: 6, aura: 2, stamina: 25, focus: 70, morale: 40 },
    captain:      { title:'Captain',      body: 5, reflex: 5, soul: 5, aura: 6, stamina: 55, focus: 55, morale: 40 },
    noble:        { title:'Noble',        body: 4, reflex: 5, soul: 6, aura: 7, stamina: 45, focus: 60, morale: 40 },
    headHunter:   { title:'Head hunter',  body: 5, reflex: 7, soul: 3, aura: 4, stamina: 60, focus: 45, morale: 40 },
    witch:        { title:'Witch',        body: 2, reflex: 5, soul: 6, aura: 2, stamina: 25, focus: 65, morale: 40 },
    valkur:       { title:'Valkur',       body: 5, reflex: 6, soul: 4, aura: 2, stamina: 60, focus: 60, morale: 40 },
    guard:        { title:'Guard',        body: 6, reflex: 5, soul: 2, aura: 4, stamina: 65, focus: 40, morale: 40 },
    gladiator:    { title:'Gladiator',    body: 7, reflex: 5, soul: 2, aura: 2, stamina: 75, focus: 25, morale: 40 },
    barbarian:    { title:'Barbarian',    body: 8, reflex: 6, soul: 2, aura: 2, stamina: 80, focus: 35, morale: 40 },
    wizard:       { title:'Wizard',       body: 1, reflex: 2, soul: 8, aura: 2, stamina: 25, focus: 80, morale: 40 },
    sage:         { title:'Sage',         body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    samurai:      { title:'Samurai',      body: 6, reflex: 7, soul: 4, aura: 5, stamina: 70, focus: 55, morale: 40 },
    ninja:        { title:'Ninja',        body: 4, reflex: 8, soul: 3, aura: 3, stamina: 55, focus: 45, morale: 40 },
    warlord:      { title:'Warlord',      body: 4, reflex: 5, soul: 6, aura: 8, stamina: 45, focus: 60, morale: 40 },
    bishop:       { title:'Bishop',       body: 4, reflex: 5, soul: 6, aura: 8, stamina: 45, focus: 60, morale: 40 },
    tracker:      { title:'Tracker',      body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    dementor:     { title:'Dementor',     body: 6, reflex: 4, soul: 1, aura: 1, stamina: 75, focus: 70, morale: 40 },
    necromanta:   { title:'Necromanta',   body: 2, reflex: 5, soul: 6, aura: 1, stamina: 45, focus: 60, morale: 40 },
    summoner:     { title:'Summoner',     body: 3, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    shaman:       { title:'Shaman',       body: 4, reflex: 5, soul: 6, aura: 6, stamina: 45, focus: 60, morale: 40 },
    thief:        { title:'Thief',        body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    rouge:        { title:'Rouge',        body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    cavalier:     { title:'Cavalier',     body: 6, reflex: 5, soul: 6, aura: 4, stamina: 45, focus: 60, morale: 40 },
    pirate:       { title:'Pirate',       body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    inquisitor:   { title:'Inquisitor',   body: 4, reflex: 5, soul: 6, aura: 4, stamina: 45, focus: 60, morale: 40 },
    icelander:    { title:'Icelander',    body: 7, reflex: 6, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    heretic:      { title:'Heretic',      body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    nomad:        { title:'Nomad',        body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    bard:         { title:'Bard',         body: 4, reflex: 5, soul: 6, aura: 6, stamina: 45, focus: 60, morale: 40 },
    warlock:      { title:'Warlock',      body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    monk:         { title:'Monk',         body: 6, reflex: 6, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    druid:        { title:'Druid',        body: 4, reflex: 5, soul: 6, aura: 4, stamina: 45, focus: 60, morale: 40 },
    illusionist:  { title:'Illusionist',  body: 3, reflex: 6, soul: 5, aura: 5, stamina: 35, focus: 50, morale: 40 },
    wardancer:    { title:'Wardancer',    body: 5, reflex: 7, soul: 6, aura: 2, stamina: 60, focus: 40, morale: 40 },
    runeSmith:    { title:'Rune smith',   body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    merchant:     { title:'Merchant',     body: 3, reflex: 3, soul: 6, aura: 6, stamina: 55, focus: 65, morale: 40 },
    explorer:     { title:'Explorer',     body: 4, reflex: 5, soul: 6, aura: 7, stamina: 45, focus: 60, morale: 40 },
    templar:      { title:'Templar',      body: 6, reflex: 3, soul: 6, aura: 5, stamina: 60, focus: 55, morale: 40 },
    spy:          { title:'Spy',          body: 4, reflex: 5, soul: 6, aura: 5, stamina: 45, focus: 60, morale: 40 },
    nun:          { title:'Nun',          body: 4, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    mercenary:    { title:'Mercenary',    body: 4, reflex: 5, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    duelist:      { title:'Duelist',      body: 3, reflex: 8, soul: 5, aura: 5, stamina: 50, focus: 60, morale: 40 },
    beastmaster:  { title:'Beastmaster',  body: 6, reflex: 5, soul: 7, aura: 1, stamina: 45, focus: 70, morale: 40 },
    sorcerer:     { title:'Sorcerer',     body: 3, reflex: 5, soul: 6, aura: 2, stamina: 45, focus: 60, morale: 40 },
    archer:       { title:'Archer',       body: 4, reflex: 7, soul: 6, aura: 3, stamina: 45, focus: 60, morale: 40 },
    psionic:      { title:'Psionic',      body: 4, reflex: 4, soul: 7, aura: 3, stamina: 45, focus: 60, morale: 40 },
    alchemist:    { title:'Alchemist',    body: 3, reflex: 5, soul: 7, aura: 2, stamina: 35, focus: 70, morale: 40 },
    enginer:      { title:'Enginer',      body: 3, reflex: 5, soul: 4, aura: 4, stamina: 35, focus: 65, morale: 40 },
    antiPaladin:  { title:'Anti Paladin', body: 7, reflex: 3, soul: 1, aura: 5, stamina: 60, focus: 60, morale: 10 },
    chaosKnight:  { title:'Chaos Knight', body: 7, reflex: 4, soul: 6, aura: 1, stamina: 50, focus: 60, morale: 10 },
    chief:        { title:'Chief',        body: 5, reflex: 5, soul: 4, aura: 7, stamina: 55, focus: 60, morale: 40 },
    maruder:      { title:'Maruder',      body: 7, reflex: 4, soul: 2, aura: 4, stamina: 60, focus: 40, morale: 40 },
    rich:         { title:'Rich',         body: 3, reflex: 4, soul: 3, aura: 8, stamina: 30, focus: 70, morale: 30 },
    shapechanger: { title:'Shapechanger', body: 6, reflex: 4, soul: 4, aura: 2, stamina: 55, focus: 45, morale: 30 },
}

// https://en.wikipedia.org/wiki/Character_class_(Dungeons_%26_Dragons)
// https://wfrp1e.fandom.com/wiki/List_Of_Advanced_Careers

export const professionList = Object.values(professionLookup);
export const professionKeyList = Object.keys(professionLookup) as ProfessionKey[];

export const traitsFactory = (level:number = 1,  professionType:ProfessionKey) => {

  const ability:Ability = professionLookup?.[professionType] ?? pickOne(professionList);

  const stamina = leveling(level, ability.stamina  / 10, 2);
  const focus    = leveling(level, ability.focus     / 10, 2);
  const morale  = leveling(level, ability.morale   / 10, 2);

  const mobTraits:MobTraits = {
    level,
    professionType,
    ability: {
      title        : ability.title,
      body         : leveling(level, ability.body     / 10, 2),
      reflex     : leveling(level, ability.reflex / 10, 2),
      soul         : leveling(level, ability.soul     / 10, 2),
      aura      : leveling(level, ability.aura  / 10, 2),
      stamina,
      focus,
      morale,
    },
    condition: {
      staminaState : stamina,
      focusState    : focus,
      moraleState  : morale,
    }
  };

  return mobTraits;
}

export const mobFactory = (name, avatar, coord, uid, team, traits:MobTraits):Mob => ({
  name,
  avatar,
  coord,
  uid,
  team,
  ...traits
});

export const increaseLevel = (amount:number = 1) => ({
  level, 
  professionType, 
  ability, 
  condition, 
  ...rest
}:Mob):Mob => ({
  ...traitsFactory(level + amount, professionType),
  ...rest
});

export const oneLevelUp = increaseLevel(1);