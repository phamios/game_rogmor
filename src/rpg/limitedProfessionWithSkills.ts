import { ProfessionKey } from "./profession";

export const limitedProfessionWithSkills:ProfessionKey[] = [
  'chaosKnight',
  'bishop',
  'icelander',
  'ninja',
  'samurai',
  'merchant',
];

export const skillForProf:Partial<Record<ProfessionKey, string[]>> = {
  'chaosKnight': ['instant target-rnd hit-soul power-1; instant target-rnd hit-body power-1; instant target-rnd hit-aura power-1','fill-3 target hit-body power-4','fill-4 target-all hit-soul power-2'],
  'bishop': ['instant target-ally heal; instant target-rnd hit-body', 'instant target hit-soul power-2','fill-3 tsa heal-2','fill-4 target-all-ally heal-3'],
  'icelander': ['instant target-all hit-body','fill-3 target-all hit-body power-2','fill-4 target-rnd hit-body power-4'],
  'ninja': ['instant target hit-body power-[2.4]','fill-4 target-all hit-body power-2','fill-2 target hit-body power-4'],
  'samurai': ['instant target hit-body power-2','fill-2 target hit-body power-4','fill-4 target-all hit-body power-2'],
  'merchant': ['instant target hit-aura; instant target-rnd hit-soul power-2','fill-2 target hit-aura power-2','fill-4 target-all hit-aura'],
};