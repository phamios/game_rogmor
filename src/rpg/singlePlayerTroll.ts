import { actionFactory, kebabToCamelCase } from 'react-troll';
import { increaseLevel, Mob } from './profession';
import { FlowAction } from './slash';

export const [getActionsLookup, action] = actionFactory(kebabToCamelCase);

export enum GameMode {
  ROLL_CHARACTER,
  ADVENTURE_ON_MAP,
}

export enum CombatOutcome {
  HERO_DIE,
  NPC_DIE,
  OVER_WITHOUT_LOSS,
}

export interface MainState {
  round : number;
  hero: Mob;
  focus: string;
  game: GameMode;
  entities: Record<string, Mob>;
  actionAnim: any;
  combatResult: any;
  damegeResult: any;
  encounterOutcome: any[];
  mobList: Mob[];
  flow?: FlowAction;
  isAutoFight: boolean;
  encounterResult: any;
}

export const initialState:MainState = {
  round : 0,
  hero: null,
  focus: null,
  game: GameMode.ROLL_CHARACTER,
  entities: {},
  actionAnim: null,
  combatResult: null,
  damegeResult: null,
  encounterOutcome: [],
  mobList: [],
  flow: null,
  isAutoFight: false,
  encounterResult: null,
};

export const 
  SET_HERO = action('set-hero'),
  MOD_HERO = action('mod-hero'),
  NEXT_ROUND = action('next-round'),
  HEART_BEAT = action('heart-beat'),
  USE_SKILL = action('use-skill'),
  SET_GAME_STATE = action('set-game-state'),
  SET_AUTO_FIGHT = action('set-auto-fight'),
  SETUP_ENTITIES = action('setup-entities'),
  MOD_ENTITI = action('mod-entiti'),
  ADD_ENTITI = action('add-entiti'),
  REMOVE_ENTITI = action('remove-entiti'),
  FOCUS_ON = action('focus-on'),
  FIGHT = action('fight'),
  SKILL = action('skill'),
  TALK  = action('talk'),
  ENCOUNTER_BEGIN  = action('encounter-begin'),
  ENCOUNTER_OUTCOME  = action('encounter-outcome'),
  ENCOUNTER_RESULT  = action('encounter-result'),
  LEVEL_UP_HERO  = action('level-up-hero'),
  USER_ACT = action("user-act"),
  PLAY_ACTION = action("play-action"),
  PLAY_OUTCOME = action("play-outcome"),
  ANIMATION_ENDED = action("animation-ended"),
  ANIMATION_SKIPPED = action("animation-skipped"),

  SET_MOB_LIST = action('set-mob-list'),
  PLAY_FLOW = action('play-flow'),

  PLAY_ACTION_ANIM = action('play-action-anim')
;

export const gameReducer = (state:MainState, {type, payload}):MainState => {
  switch (type) {
    case SET_HERO: return {...state, hero: payload};
    case MOD_HERO: return {...state, hero: payload(state.hero), combatResult:null};
    case NEXT_ROUND: return {...state, round: payload(state.round)};
    case SET_GAME_STATE: return {...state, game:payload};
    case SETUP_ENTITIES: return {...state, entities: payload};
    case SET_MOB_LIST: return {...state, mobList: payload};
    case MOD_ENTITI: {
      const {entities} = state;
      return entities[payload?.uid] 
        ? {...state, entities: {...entities, [payload?.uid]: payload} }
        : state;
    };
    case FOCUS_ON: return {...state, focus: payload};
    case PLAY_ACTION_ANIM: return {...state, actionAnim: payload};
    case PLAY_FLOW: return {...state, flow: payload};
    case ENCOUNTER_BEGIN: return {...state, encounterResult: null};
    case ENCOUNTER_RESULT: return {...state, encounterResult: payload};
    case SET_AUTO_FIGHT: return {...state, isAutoFight: payload};

    case LEVEL_UP_HERO: return {...state, hero: increaseLevel(1)(state.hero)}
    default: return state;
  }
};

const lostFocus = s => s

const checkIsLive = ({condition}:Mob) => (
     condition.staminaState > 0 
  && condition.focusState > 0 
  && condition.moraleState > 0
);