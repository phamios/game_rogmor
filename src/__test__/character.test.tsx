import { oneLevelUp, increaseLevel, Mob, mobFactory, Team, traitsFactory, professionKeyList } from "../rpg/profession";
import { improved, rnd, amount } from "../rpg/rpg";

const fighter1Trait = traitsFactory(1, 'fighter');
const fighterMob:Mob = {...fighter1Trait, 
  name: 'Mr. Foo',
  coord: 0,
  avatar: 22,
  uid: 'dfx532',
  team: Team.BAD,
};

const actionOrder = (mobList:Mob[]) => mobList
  .map(mob => [mob, improved(
      (
        mob.ability.reflex 
      + (mob.condition.staminaState / 10)
      + (mob.condition.focusState / 10)
      ) / mob.level
    )])
  .sort(([a, aSpeed],[b, bSpeed]) => aSpeed > bSpeed ? -1 : 1 )
;

test ('create lvl 1 fighter', () => {  
  expect(fighter1Trait).toMatchSnapshot();
});


test ('Personalize lvl 1 fighter', () => {
  expect(fighterMob).toMatchSnapshot();
});

test ('Standard Personalize lvl 1 fighter', () => {
  const mobBymobFactory:Mob = mobFactory('Mr. Foo', 22, 0, 'dfx532', Team.BAD, fighter1Trait)
  expect(mobBymobFactory).toStrictEqual(fighterMob); 
});

test ('Level Up to fighter 2', () => {
  const MrFooLevel2 = increaseLevel(1)(fighterMob);
  expect(MrFooLevel2).toMatchSnapshot();
});

test ('Level Up to fighter 2 standard', () => {
  const mrFooLevel2 = increaseLevel(1)(fighterMob);
  const mrFooLevel2v2 = oneLevelUp(fighterMob);
  expect(mrFooLevel2).toStrictEqual(mrFooLevel2v2);
});

test('order of speed', () => {
  const MrFooLevel2 = increaseLevel(10)(fighterMob);
  const list = [fighterMob, MrFooLevel2];
  const order = actionOrder(list);
  
  expect(order).toEqual([[MrFooLevel2, 20], [fighterMob, 4]])
})

test('speed order of different profession lvl 10', () => {
  const list:Mob[] = professionKeyList.map(
    (key, index) => mobFactory(
      `Mr. ${key}`,7,0,index,Team.GOOD,
      traitsFactory(100, key)
    )
  )

  const order = actionOrder(list);
  
  expect(
    order.map(([mob, speed]:[Mob, number]) => `${mob.professionType}`)
  ).toMatchSnapshot();
})

test('Samurai speed number by levels', () => {
  const makeSamurai = (lvl) => 
    mobFactory(
      `Samurai level:${lvl}`,1,2,`id:${lvl}`,Team.GOOD,
      traitsFactory(lvl, 'samurai')
    )
  ;

  const samuraiList = amount(20).map(lvl => makeSamurai((lvl + 1) * 5));

  expect(
    actionOrder(samuraiList).map(
      ([mob, speed]:[Mob, number]) => `${mob.name} ${speed}`
    )
  ).toMatchSnapshot();
})

test('Fighter speed number by levels', () => {
  const makeFighter = (lvl) => 
    mobFactory(
      `Fighter level:${lvl}`,1,2,`id:${lvl}`,Team.GOOD,
      traitsFactory(lvl, 'fighter')
    )
  ;

  const fighterList = amount(20).map(lvl => makeFighter((lvl + 1) * 5));

  expect(
    actionOrder(fighterList).map(
      ([mob, speed]:[Mob, number]) => `${mob.name} ${speed}`
    )
  ).toMatchSnapshot();
})

