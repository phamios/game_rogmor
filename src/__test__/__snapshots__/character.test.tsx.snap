// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fighter speed number by levels 1`] = `
Array [
  "Fighter level:100 163",
  "Fighter level:95 155",
  "Fighter level:90 147",
  "Fighter level:85 139",
  "Fighter level:80 131",
  "Fighter level:75 123",
  "Fighter level:70 115",
  "Fighter level:65 107",
  "Fighter level:60 99",
  "Fighter level:55 91",
  "Fighter level:50 83",
  "Fighter level:45 75",
  "Fighter level:40 67",
  "Fighter level:35 59",
  "Fighter level:30 51",
  "Fighter level:25 43",
  "Fighter level:20 34",
  "Fighter level:15 26",
  "Fighter level:10 18",
  "Fighter level:5 10",
]
`;

exports[`Level Up to fighter 2 1`] = `
Object {
  "ability": Object {
    "aura": 6,
    "body": 7,
    "focus": 22,
    "morale": 20,
    "reflex": 6,
    "soul": 5,
    "stamina": 26,
    "title": "Fighter",
  },
  "avatar": 22,
  "condition": Object {
    "focusState": 22,
    "moraleState": 20,
    "staminaState": 26,
  },
  "coord": 0,
  "level": 2,
  "name": "Mr. Foo",
  "professionType": "fighter",
  "team": 2,
  "uid": "dfx532",
}
`;

exports[`Personalize lvl 1 fighter 1`] = `
Object {
  "ability": Object {
    "aura": 2,
    "body": 3,
    "focus": 7,
    "morale": 6,
    "reflex": 2,
    "soul": 2,
    "stamina": 8,
    "title": "Fighter",
  },
  "avatar": 22,
  "condition": Object {
    "focusState": 7,
    "moraleState": 6,
    "staminaState": 8,
  },
  "coord": 0,
  "level": 1,
  "name": "Mr. Foo",
  "professionType": "fighter",
  "team": 2,
  "uid": "dfx532",
}
`;

exports[`Samurai speed number by levels 1`] = `
Array [
  "Samurai level:100 227",
  "Samurai level:95 215",
  "Samurai level:90 204",
  "Samurai level:85 193",
  "Samurai level:80 182",
  "Samurai level:75 170",
  "Samurai level:70 159",
  "Samurai level:65 148",
  "Samurai level:60 137",
  "Samurai level:55 126",
  "Samurai level:50 114",
  "Samurai level:45 103",
  "Samurai level:40 92",
  "Samurai level:35 81",
  "Samurai level:30 70",
  "Samurai level:25 58",
  "Samurai level:20 47",
  "Samurai level:15 36",
  "Samurai level:10 25",
  "Samurai level:5 14",
]
`;

exports[`create lvl 1 fighter 1`] = `
Object {
  "ability": Object {
    "aura": 2,
    "body": 3,
    "focus": 7,
    "morale": 6,
    "reflex": 2,
    "soul": 2,
    "stamina": 8,
    "title": "Fighter",
  },
  "condition": Object {
    "focusState": 7,
    "moraleState": 6,
    "staminaState": 8,
  },
  "level": 1,
  "professionType": "fighter",
}
`;

exports[`speed order of different profession lvl 10 1`] = `
Array [
  "samurai",
  "duelist",
  "dementor",
  "valkur",
  "ninja",
  "headHunter",
  "barbarian",
  "archer",
  "assasin",
  "wardancer",
  "icelander",
  "monk",
  "beastmaster",
  "chief",
  "captain",
  "rogue",
  "noble",
  "guard",
  "sage",
  "warlord",
  "bishop",
  "tracker",
  "necromanta",
  "summoner",
  "shaman",
  "thief",
  "rouge",
  "cavalier",
  "pirate",
  "inquisitor",
  "heretic",
  "nomad",
  "bard",
  "warlock",
  "druid",
  "runeSmith",
  "explorer",
  "spy",
  "nun",
  "mercenary",
  "sorcerer",
  "alchemist",
  "gladiator",
  "merchant",
  "enginer",
  "antiPaladin",
  "chaosKnight",
  "illusionist",
  "templar",
  "psionic",
  "fighter",
  "knight",
  "witch",
  "maruder",
  "rich",
  "shapechanger",
  "mage",
  "priest",
  "paladin",
  "wizard",
]
`;
